### PR TITLE
add -conf=path compiler switch

### PR DESCRIPTION
**Pull request by MartinNowak**

### CODE REVIEW

**Comment by MartinNowak** [comment]
See references for how people struggle with the current behavior.

**Comment by WalterBright** [comment]
People tend to "fire and forget" their dmd.conf. Changing the search order will unexpectedly break things for them. It's like changing the way `make` works - what a disaster that is for people who cut&paste their makefiles with no understanding.

**Comment by yebblies** [comment]
> This order makes more sense, because you want to configure different
compilers so the configuration should be attached to the compiler not
your current working dir.

On the other hand, sometimes you want to override sc.ini for a local project.  I'm not sure what the best way is, but it's really annoying not being able to use a specific config file.

**Comment by MartinNowak** [comment]
> A reasonable case can be made for just about any ordering.

No it can't, it's an installation time configuration for the compiler. That doesn't change when you cd into another directory. The current search order is nonsense because it favors using the same configuration for multiple compilers over using specific configurations per compiler.

> On the other hand, sometimes you want to override sc.ini for a local project.

That's what you'd use a compiler switch for that specifies an alternative configuration.
Picking up the conf of the current directory means it will be used by any different compiler which leads to the know trouble.

> People tend to "fire and forget" their dmd.conf. Changing the search order will unexpectedly break things for them.

Yeah, I'm also worried about that. Now I wonder who uses the current search order to override it's configuration. We could detect that and ask the people to move the config files.
And we should still add a switch to the compiler to select a specific config.

**Comment by yebblies** [comment]
> That's what you'd use a compiler switch for that specifies an alternative configuration.

Sure, if we had one.

> And we should still add a switch to the compiler to select a specific config.

Yes please.

**Comment by AndrejMitrovic** [comment]
> And we should still add a switch to the compiler to select a specific config.

That would probably be the best solution.

**Comment by MartinNowak** [comment]
I'm search for both the new and the old order. If there is a mismatch an informational warning is printed, advising people to use the newly added `-conf=path` switch to select a non-standard configuration.

Once again the old order is culprit for many dmd.conf related issues, because it doesn't make sense that my installed compiler picks up a different config file depending on where I am.
Note that it's still possible to override the config of the system wide dmd by adding a dmd.conf in your HOME folder.
The current working directory will only be used as last resort, because you do __not__ want to configure all compilers on a per project base.

**Comment by WalterBright** [comment]
> because it doesn't make sense that my installed compiler picks up a different config file depending on where I am.

It seems obvious to me that when I am in my project directory, that it picks up the config file from that directory. I use it all the time.

Anyhow, a compiler flag sounds fine, but not a redo of the order and the attendant confusion and breakage that will cause.

> If there is a mismatch an informational warning is printed

That just seems like layers of more complexity.

**Comment by andralex** [comment]
I disagree with the proposed order; local stuff should override global stuff. So things are good as they are. To override the global stuff just create a local conf file.

I'm okay with adding a flag though I see all this unnecessary churn and the illusion of progress because a simple and effective solution is available.

**Comment by MartinNowak** [comment]
> I disagree with the proposed order; local stuff should override global stuff.

Yes, local to the compiler you're using.
Anyhow can't seem to convince anyone, so I'll just add the flag.

**Comment by MartinNowak** [comment]
Now it's only the -conf=path switch. I also left in the additions to error.c.

> It seems obvious to me that when I am in my project directory, that it picks up the config file from that directory. I use it all the time.

It's obvious for things that you want to configure on a per-project base, like say which compiler version to use. But it doesn't make sense to use the same compiler configuration for every compiler.
And because `@P` is relative to the dmd.conf path, you'd have to hardcode the path to phobos and druntime. How is that useful? It makes it impossible to compile the project with different dmd compilers.
And what is it that you're actually configuring? DFLAGS?

File: src/mars.c [hunk 1]
```diff
@@ -1808,6 +1826,27 @@ static const char* parse_arch(size_t argc, const char** argv, const char* arch)
...
+{
+    const char *conf=NULL;
+    for (size_t i = 0; i < argc; ++i)
+    {
+        const char* p = argv[i];
+        if (p[0] == '-')
+        {
+            if (strncmp(p + 1, "conf=", 5) == 0)
+                conf = p + 6;
+            else if (strcmp(p + 1, "run") == 0)
```
Comment (andralex) [inline src/mars.c]: what's "run" doing here?

**Comment by andralex** [comment]
If `-conf=` is passed without an argument, will that make dmd to use no conf file at all?

File: src/errors.c [hunk 2]
```diff
@@ -136,15 +144,11 @@ void warning(Loc loc, const char *format, ...)
     va_end( ap );
 }
 
-/**************************************
- * Print supplementary message about the last error
- * Used for backtraces, etc
- */
-void errorSupplemental(Loc loc, const char *format, ...)
+void warningSupplemental(Loc loc, const char *format, ...)
```
Comment (yebblies) [inline src/errors.c]: Don't add these if they're not used.

Comment (MartinNowak) [inline src/errors.c]: I already needed them earlier and it just makes the API consistent.

File: src/mars.c (hunk 1 above)
Comment (MartinNowak) [inline src/mars.c]: In `dmd -run myprog.d -conf=dmd.conf` the `-conf=dmd.conf` argument belongs to myprog.

**Comment by MartinNowak** [comment]
> If -conf= is passed without an argument, will that make dmd to use no conf file at all?

Good idea, added that.

File: src/inifile.c [hunk 3]
```diff
@@ -296,7 +299,7 @@ const char *inifile(const char *argv0, const char *inifile, const char *envsecti
                 break;
         }
     }
-    return filename;
+    return;
```
Comment (andralex) [inline src/inifile.c]: unnecessary return at end of function

**Comment by andralex** [comment]
Auto-merge toggled on

**Comment by MartinNowak** [comment]
tl;dr, please reconsider changing the conf order or splitting the conf file

That dmd.conf is driving me crazy.

I need a dmd.conf in my dmd repo, so that I can use dmd-master. It wouldn't work without a config and putting `dmd.conf` anywhere but near dmd itself overrides my system dmd.conf.
That setup worked nicely for me although some of you guys seem to have a dmd.conf in your home dir (why?).

Now that we require a host D compiler to build dmd it no longer works, because the dmd.conf for my dmd-master overrides the system wide dmd.conf of my host dmd.
Why would my system dmd need a different configuration just because __I__ am in a different folder?
It doesn't, it still uses the same phobos (the system wide) and still requires the same linker switches.

Setting HOST_DC is a workaround but is impractical, because it requires an awkward `env HOST_DC='dmd -conf=/etc/dmd.conf' make -f posix.mak` to work.
Requiring `dmd -conf=/etc/dmd.conf` in order to not pickup some random conf file is crazy. It's not portable either, because the config is somewhere else on each platform.

We learned the same from dlang.org https://github.com/D-Programming-Language/dlang.org/pull/758#issuecomment-74294012, adding the `-conf=` switch didn't solve the actual problem and created new ones.

It might seem intuitive, that a local configuration should override a global one, but it's a false friend here, because the configuration should be local from the perspective of the compiler, not from the perspective of the user.

In fact the config is so tied to the compiler that we could almost compile it __into__ the compiler (like gcc does with it's spec). At best it's something package maintainers need to touch to accomodate platform differences, but it's not a user configuration file.

If people use it to configure DFLAGS and such project dependent stuff, then we need to split the configuration file, into one part that tells the compiler where to find druntime/phobos and how to link, and one part that can be used for per-project configuration of compiler arguments and env variables.
IMO the latter is better kept in makefiles/dub.json though.

### conclusion

The lookup order for the config file should be changed to the following.
- next to dmd binary (highest precedence so that I can have multiple installations)
- per-user conf folder (HOME) (to override the system-wide config)
- system-wide conf folder (to allow package installations .deb/.rpm)

The current situation is unmaintainable.
